### DISCUSSION TITLE
Integrate sound boost (if possible?)

### DISCUSSION THREAD

**Discussion by Jeremy2131** [radj307/volume-control#84] [Ideas & Feature Discussions]
Title explains my idea, sometimes for me personally I have a very low volume videos on youtube or streams and yes I have installed a plugin to boost the sound but it requires me to do additional clicks when I could just do it with volume control, if it won't take much time I think it would be a good feature to have

**Comment by radj307** [comment]
I'll look into this and see what I can find!

Edit:  
I'm confused by what you mean by a plugin? Is this 3rd party software, a browser extension, or through `mmsys.cpl` (sound settings)?:  
![image](https://github.com/radj307/volume-control/assets/1927798/d8476f49-a5a2-4f43-b86d-2c42291973ce)

**Comment by radj307** [comment]
The boost level that appears in `mmsys.cpl` is provided by the audio device driver and isn't exposed directly through the Core Audio API. That doesn't mean it's impossible to access (after all, the sound settings have to do it somehow), but it does mean it'll take a lot more research before I can come up with a way to implement this.

**Comment by radj307** [comment]
resources for further reasearch:  
- https://learn.microsoft.com/en-us/windows/win32/coreaudio/core-audio-interfaces
- https://learn.microsoft.com/en-us/windows/win32/coreaudio/audio-endpoint-properties
- https://learn.microsoft.com/en-us/windows/win32/coreaudio/pkey-audioendpoint-controlpanelpageprovider (CLSID of the provider for device control panel settings in `mmsys.cpl`)
- https://stackoverflow.com/questions/31132881/how-to-set-microphone-boost-in-c-sharp-using-wasapi
- https://github.com/microsoft/Windows-driver-samples/tree/main/audio/sysvad (sample implementation of an audio device driver)

**Comment by radj307** [comment]
ChatGPT confirmed my thoughts on this, but also pointed me in the direction of the `SetupDi` family of functions that would allow access to device driver properties. Accessing the gain itself may be an issue as it varies between drivers.  
It may be possible to somewhat reliably find the property that controls gain by using regex to search for keywords like "gain" or "boost". This would almost certainly require administrator permissions to access, though. Doing this could cause the application to be flagged as malware, so if I do this it'll probably be as an optional addon that isn't distributed with the main application.

additional resources:
- https://learn.microsoft.com/en-us/windows-hardware/drivers/install/overview-of-device-interface-classes
- https://learn.microsoft.com/en-us/windows-hardware/drivers/install/using-a-device-interface
- https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nf-setupapi-setupdienumdeviceinterfaces

**Comment by radj307** [comment]
As an alternative, I could implement a custom volume booster to increase the audio gain for playback devices regardless of whether they actually provide it, similar to how other 3rd party volume booster applications work.

I have wanted to implement hotkeys that could interact with EqualizerAPO in some way, which would be cool.

**Comment by radj307** [comment]
This appears to be what I was looking for:

- https://learn.microsoft.com/en-us/windows/win32/api/devicetopology/nn-devicetopology-iaudiovolumelevel

**Reply by Jeremy2131** [reply]
Looks like that, yes

**Comment by Jeremy2131** [comment]
I personally use this web extenstion for chrome: https://chrome.google.com/webstore/detail/sound-booster-that-works/gnidjfdekbljleajoeamecfijnhbgndl
But it would be cool to have something similar for all apps in Windows
